plugins/usbdmx/DMXCreator512Basic.cpp

```diff
@@ -0,0 +1,287 @@
+    : DMXCreator512Basic(adaptor, usb_device, serial) {
+}
+
+bool SynchronousDMXCreator512Basic::Init() {
+  libusb_device_handle *usb_handle;
+
+  bool ok = m_adaptor->OpenDeviceAndClaimInterface(
+      m_usb_device, 0, &usb_handle);
+  if (!ok) {
+    return false;
+  }
+
+  std::auto_ptr<DMXCreator512BasicThreadedSender> sender(
+      new DMXCreator512BasicThreadedSender(m_adaptor, m_usb_device,
+                                           usb_handle));
+  if (!sender->Start()) {
+    return false;
+  }
+  m_sender.reset(sender.release());
+  return true;
+}
+
+bool SynchronousDMXCreator512Basic::SendDMX(const DmxBuffer &buffer) {
+  return m_sender.get() ? m_sender->SendDMX(buffer) : false;
+}
+
+// DMXCreator512BasicAsyncUsbSender
+// -----------------------------------------------------------------------------
+class DMXCreator512BasicAsyncUsbSender : public AsyncUsbSender {
+ public:
+  DMXCreator512BasicAsyncUsbSender(LibUsbAdaptor *adaptor,
+                                   libusb_device *usb_device)
+      : AsyncUsbSender(adaptor, usb_device),
+        m_adaptor(adaptor),
+        m_usb_device(usb_device) {
+    m_dmx_buffer_1 = new uint8_t[CHANNELS_PER_PACKET];
+    m_dmx_buffer_2 = new uint8_t[CHANNELS_PER_PACKET];
+    m_state = STATE_SEND_STATUS;
+  }
+
+  ~DMXCreator512BasicAsyncUsbSender() {
+    CancelTransfer();
+    delete[] m_dmx_buffer_1;
+    delete[] m_dmx_buffer_2;
+  }
+
+  libusb_device_handle* SetupHandle() {
+    libusb_device_handle *usb_handle;
+    bool ok = m_adaptor->OpenDeviceAndClaimInterface(
+        m_usb_device, 0, &usb_handle);
+    return ok ? usb_handle : NULL;
+  }
+
+  bool PerformTransfer(const DmxBuffer &buffer) {
+    unsigned int length = CHANNELS_PER_PACKET;
+    memset(m_dmx_buffer_1, 0, length);
+    buffer.Get(m_dmx_buffer_1, &length);
+
+    length = CHANNELS_PER_PACKET;
+    memset(m_dmx_buffer_2, 0, length);
+    buffer.GetRange(CHANNELS_PER_PACKET, m_dmx_buffer_2, &length);
+
+    m_state = STATE_SEND_FIRST_HALF;
+    FillBulkTransfer(ENDPOINT_1,
+                     const_cast<unsigned char*>(status_buffer),
+                     sizeof(status_buffer), URB_TIMEOUT_MS);
+    return (SubmitTransfer() == 0);
+  }
+
+  void PostTransferHook() {
+    switch (m_state) {
+      case STATE_SEND_STATUS:
+        // handled in PerformTransfer()
+        break;
+      case STATE_SEND_FIRST_HALF:
+        m_state = STATE_SEND_SECOND_HALF;
+        FillBulkTransfer(ENDPOINT_2,
+                         const_cast<unsigned char*>(m_dmx_buffer_1),
+                         CHANNELS_PER_PACKET, URB_TIMEOUT_MS);
+        SubmitTransfer();
+        break;
+      case STATE_SEND_SECOND_HALF:
+        m_state = STATE_SEND_STATUS;
+        FillBulkTransfer(ENDPOINT_2,
+                         const_cast<unsigned char*>(m_dmx_buffer_2),
+                         CHANNELS_PER_PACKET, URB_TIMEOUT_MS);
+        SubmitTransfer();
+        break;
+    }
+  }
+
+
+ private:
+  uint8_t *m_dmx_buffer_1;
```

Comment:
same comment here, make these arrays instead of pointers - at least in this case we're not leaking them.